dev: add 'generate js' subcommand

sjbarag opened the pull request:
Interactive development of UI features requires a few prerequisites to
be built to satisfy editor requirements. Specifically, cluster-ui
depends on the typings produced from building
@cockroachlabs/crdb-protobuf-client, db-console depends on the typings
produced from building both the protobuf client and cluster-ui, and
eslint editor plugins require @cockroachlabs/eslint-plugin-crdb to be
compiled and the compiled output to be in the Bazel workspace. Add a
./dev generate js target that compiles the minimum set of dependencies
to bootstrap a development environment.

This requires a small change to the `tsconfig.json` for cluster-ui and
db-console to ensure they know where to find the Bazel-produced output
of the packages in this repo.

fixes #84738

sjbarag committed:
dev: delete eslint plugin and db-console build artifacts with dev ui clean

./dev ui clean previously only removed the build output of cluster-ui
and the JS protobuf client, accidentally leaving old versions of
eslint-plugin-crdb and db-console around in the Bazel workspace. Remove
those build artifacts as well to ensure a clean local state.

Release note: None

sjbarag committed:
dev: add ./dev generate js subcommand

Interactive development of UI features requires a few prerequisites to
be built to satisfy editor requirements. Specifically, cluster-ui
depends on the typings produced from building
@cockroachlabs/crdb-protobuf-client, db-console depends on the typings
produced from building both the protobuf client and cluster-ui, and
eslint editor plugins require @cockroachlabs/eslint-plugin-crdb to be
compiled and the compiled output to be in the Bazel workspace. Add a
./dev generate js target that compiles the minimum set of dependencies
to bootstrap a development environment.

fixes #84738

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84766)
<!-- Reviewable:end -->

sjbarag committed:
ui: teach tsc about bazel build output

By default, the TypeScript compiler expects dependency typings to be in
./node_modules/:dependency_name/. Unfortunately, the types produced for
packages contained in this repo (e.g.
@cockroachlabs/crdb-protobuf-client and @cockroachlabs/cluster-ui), are
placed in /path/to/cockroach.git/_bazel/bin/pkg/ui/workspaces/... .
Building, linting, and testing assets with Bazel all work successfully,
but interactive development previously required files to be moved around
or assets to be rebuilt via make. Use TypeScript's paths compiler option
[1] to read typings for cluster-ui and crdb-protobuf-client from both
the make and Bazel output directories.

[1] https://www.typescriptlang.org/tsconfig#paths

Release note: None

sjbarag commented:
Based on #84856 , so ignore the first commit :)

sjbarag commented:
I broke the ui tests, checking that out now

rickystewart commented on pkg/cmd/dev/generate.go:
```diff
@@ -323,3 +325,41 @@ import "C"
+	}
+	logCommand("bazel", args...)
+	if err := d.exec.CommandContextInheritingStdStreams(ctx, "bazel", args...); err != nil {
+		return fmt.Errorf("Building JS development prerequisites: %s", err.Error())
```
This is throwing a lint error, but you should just be able to do

```
return fmt.Errorf("building JS development prerequisites: %w", err)
```

Note the sentence should start with a lowercase letter.

sjbarag replied:
How strange, I ran `./dev lint` and everything passed 0.o

rickystewart replied:
Ah sorry, that will only build `cockroach` for lints, not `dev`. `bazel build pkg/cmd/dev` will catch this.

rickystewart commented on pkg/cmd/dev/generate.go:
```diff
@@ -323,3 +325,41 @@ import "C"
+func (d *dev) generateJs(cmd *cobra.Command) error {
+	ctx := cmd.Context()
+
+	args := []string{
```
Looks like `dev generate js` doesn't call into `copyBuildOutputToWorkspace` and just does a `CopyAll`. Is that right? Is there any particular reason why it is not necessary here?

sjbarag replied:
It is right!  The original idea was to use `copyBuildOutputToWorkspace()` to copy all of the compiled typings back into the bazel workspace, but that'd require someone to rerun `./dev generate js` after every build for their IDE to see those changes.  That's a pretty tall order to ask of folks, so I made all of the TypeScript IDE tooling look in `_bazel/bin/…` for `.d.ts` files in addition to the normal `./node_modules/` location.  This allows a `./dev build` to produce new typings and for IDEs to pick them up immediately.  The only thing that did need to get copied into the workspace is a custom plugin for the JS linter: it has to be compiled before it can be used, and `eslint` doesn't seem to expose a configuration option to expand the search locations for plugins beyond the normal `node_modules` patterns.

rickystewart commented:
LGTM, but it seems like the Jest tests are failing?

sjbarag commented:
> LGTM, but it seems like the Jest tests are failing?

Yep, I've got a fix for that coming very soon :)

rickystewart reviewed (approved):
Nice, thanks Sean :)

sjbarag commented:
bors r=rickystewart
